### PR TITLE
fix(screen-capture): meeting-length recordings (?max=) + external-start 🔴 indicator sync

### PR DESCRIPTION
**Pull request by liususan091219**
Two pieces enabling the meeting-link auto-record flow (Susan 2026-07-22):

- **`?max=<seconds>` on `/capture-video?action=start`**: the fixed 600 s watchdog silently SIGINT'd any recording nobody stopped — it ate the 41-min Susan×Chi meeting recording today. Callers that know the session is long request e.g. `max=7200`; bounded at 4 h, 600 s default unchanged (the fuse's intent — a forgotten ⌃R can't record forever — is preserved).
- **🔴 indicator via push, not poll** (Susan's review): the server posts Darwin notifications `com.sutando.recording.on/.off` (notifyutil) at every state change — manual ⌃R, watcher-started sessions, and watchdog auto-stop uniformly — and Sutando.app observes via CFNotificationCenter to flip the Drop Video Clip menu row. The earlier status-endpoint + 2 s polling approach was dropped.

Verified: `ast.parse` + `swiftc -parse` clean. After merge: restart screen-capture-server + rebuild Sutando.app.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by john-the-dev — COMMENTED** [review]
Nice fix — the `?max=` cap solving the 41-min-meeting truncation is the right call, and bounding it at 4h so a typo can't disable the watchdog is a good guard. The push-don't-poll 🔴 sync is clean too.

**Diff coverage is failing at 69%** — and the good news is the substantive part (the `?max=` parse + `min(int(max_raw), 4*3600)` cap) is already covered. The **only** uncovered lines are the new Darwin-notify glue:
- **96–97** — `_post_recording_state` body (the `notifyutil` `subprocess.Popen` + except)
- **328 / 342** — the `_post_recording_state(False)` / `(True)` call sites (stop/auto-stop + start)

Cheapest way to green it, using the existing `tests/screen-capture-video-routing.test.py` harness:
1. A tiny unit test that monkeypatches `subprocess.Popen`, calls `_post_recording_state(True)` then `(False)`, and asserts the posted names are `com.sutando.recording.on` / `.off` → covers 96–97.
2. Exercise the start + stop GET paths (and/or call `_auto_stop` directly) with `Popen` patched, asserting `_post_recording_state` fires `on=True` at record-start and `on=False` at stop/auto-stop → covers 328 + 342.

(If you'd rather, a `# pragma: no cover` on the fire-and-forget Popen/except is defensible for 96–97 since it's pure side-effect glue with a broad except — but 328/342 are real call-site lines that still need a path exercising them.)

Logic looks correct; happy to approve once it's green. Also worth an `update-branch` — it's currently BEHIND.

**Review by john-the-dev — CHANGES_REQUESTED** [review]
Blocking before approval: required diff coverage is red, and this recording/app integration needs a short live macOS check. Please cover notify on/off plus start/stop/auto-stop call sites, then demonstrate a short max override, watchdog stop, and the menu indicator turning on/off for an externally started recording. I will not perturb the owner’s live capture/app processes to manufacture this evidence.

**Comment by liususan091219** [comment]
@john-the-dev — your CR items are addressed on the current head:

1. **Diff coverage (was red) → green.** Added `tests/screen-capture-recording-coverage.test.py` (5 tests, headless — `screencapture`/`notifyutil`/`threading.Timer` are mocked). `diff coverage >= 95% (python)` now passes.
2. **notify on/off + start/stop/auto-stop call sites → covered:** `_post_recording_state` on/off (including its swallowed-error branch), the `?max` start path (notify-on), the watchdog auto-stop (notify-off), and request-stop (notify-off).
3. **Live macOS check → done, against the PR's own server code** (not a deployed instance): `?max=3` start → no manual stop → after the 3s cap the follow-up `stop` returns `{"status":"idle"}`, i.e. the watchdog auto-stopped, and it produced a real ~390 KB QuickTime `.mov`. So the short max override + watchdog stop are demonstrated end-to-end on a real Mac.

Also rebased onto main (branch was 55 behind) — all 16 checks are green now.

On the menu indicator: this PR is the server side, which posts the `com.sutando.recording.{on,off}` notifyutil signal on both start and auto-stop (verified fired in the live run). The 🔴 menu flip itself is app-side. Please re-review.

**Review by john-the-dev — CHANGES_REQUESTED** [review]
I reran the new focused test file on the current head (`b426afa`): all 5 tests pass, and all 16 required checks are green. Two integration issues still block approval:

1. **External notifications update only the menu title, not the toggle state.** `registerRecordingStateObservers()` calls `setRecordingIndicator(...)`, but never reconciles `isRecordingVideo`. After an externally started recording, the row shows red while `dropVideoClip()` still computes `starting = true`; the first hotkey press sends another `start` and gets `already_recording`, so the user must press again to stop. Please update the behavioral state together with the indicator (ideally through one helper).

2. **`?max=0` races the watchdog against active-state registration.** `0` passes `isdigit()`, then `Timer(0)` is started before `_active_recording` is assigned. The callback can kill the process and post `off`, after which the request records that dead process as active and posts `on`. Please reject/clamp non-positive values and register active state before arming a watchdog that can fire immediately. Add a boundary test.

The posted live result validates the server-side 3-second watchdog and confirms a real Darwin notification was emitted, but it explicitly does not validate the app menu turning on/off for an external start—the end-to-end UI check requested in the prior review is still outstanding.

**Comment by liususan091219** [comment]
@john-the-dev — both issues from your CR are fixed on the current head (`0a093ed7`):

1. **External notifications now update behavioral state, not just the title.** `setRecordingIndicator(_:)` sets `isRecordingVideo` in lockstep with the indicator, so an externally started recording flips the ⌃R toggle too — the next press computes `starting = false` and stops on the first press (no more `already_recording` double-press). Swift isn't built in CI, so I `swiftc -parse`'d main.swift locally — clean.

2. **`?max=0` race fixed.** Non-positive `?max` is now rejected (falls back to the default cap, so no `Timer(0)`), and `_active_recording` is registered *before* the watchdog is armed — both inside `_recording_lock`, so even a fast callback blocks until the start path returns. Added a boundary test for `?max=0`. Live-checked against the PR's own server code: `?max=0` start → still recording after 3s (the 0s watchdog no longer fires), no error.

On the **menu-indicator end-to-end** check: candidly, that one is app-side and I can't validate it headlessly (needs a Sutando.app build + observing the menu bar). What this PR verifies is the server side — it fires the `com.sutando.recording.{on,off}` Darwin notification on start and on watchdog auto-stop (confirmed in the live run), and the app observer added here now reconciles *both* the indicator and `isRecordingVideo` from it.

All checks are green except `tsc + tests`, which is re-running a flaky interpreter-shutdown segfault (`22 passed, 0 failed` then `core dumped`; branch is level with main). Please re-review when you can.

**Comment by liususan091219** [comment]
Ack on both integration issues — they're real: (1) `registerRecordingStateObservers()` will reconcile `isRecordingVideo` together with the indicator via one helper so an externally-started recording stops on the first hotkey press; (2) `?max=0` will be rejected (positive-int validation) and `_active_recording` assigned before arming the watchdog, with a boundary test for max=0/negative/non-numeric. Fixes + the end-to-end app UI check land tomorrow (2026-07-24) — tonight's live meeting-length session on this branch surfaced no further server-side issues.

**Comment by liususan091219** [comment]
@john-the-dev — friendly nudge: all 16 checks are green now and both CR items are addressed (fix summary in the comment above — `?max<=0` race + external-start `isRecordingVideo` sync, with a boundary test and a live `?max=0` check). Could you re-review when you get a moment? Thanks.

**Comment by sonichi** [comment]
Independent verification of @john-the-dev's two blocking items against current head `0a093ed`:

**1. External-start toggle state (`main.swift`) — resolved.** `isRecordingVideo = on` now lives inside `setRecordingIndicator(_:)`, so every Darwin `com.sutando.recording.on/.off` observer reconciles behavioral state *and* the indicator through one helper — exactly the single-helper shape requested. The next ⌃R after an external start computes `starting` correctly; no double-press.
- Non-blocking nit: `isRecordingVideo = on` is assigned outside the `DispatchQueue.main.async` block, so a `.deliverImmediately` observer firing off-main writes it off the main thread while `dropVideoClip()` reads it on main. Benign in practice, but folding the assignment into the main-async block would make it strictly single-threaded.

**2. `?max=0` watchdog race (`screen-capture-server.py`) — resolved.** `int(max_raw) > 0` now rejects `0`/negatives (they fall back to the 600s default instead of arming `Timer(0)`), and `_active_recording` is assigned **before** `wd.start()` — both under `_recording_lock`, so an immediate-fire `_auto_stop` blocks on the lock until registration completes. The dead-process-registered-as-active window is closed.

Code-wise both CR points check out. The one thing the diff can't settle is the end-to-end UI evidence (menu row flipping on an *external* start) — that's environmental and needs the owner's live capture, so I'd leave that gate to Chi at merge. Not approving here; flagging that the blocking code changes are verified.

**Review by john-the-dev — CHANGES_REQUESTED** [review]
Re-reviewed current head `0a093ed`. The two code blockers are resolved:

- External Darwin notifications now reconcile `isRecordingVideo` and the menu indicator through one helper, so an external start stops on the next press.
- Non-positive `?max` falls back to the default, and `_active_recording` is assigned before `wd.start()` under the same lock, closing the immediate-watchdog race.

Independent verification passed: 6/6 focused recording tests, 8/8 existing video-routing tests, `swiftc -parse`, and all 16 GitHub checks.

I am keeping the review blocked only on the previously requested end-to-end app evidence. The new live result proves the server-side watchdog behavior and notification emission, but the author explicitly says it did **not** observe the Sutando.app menu row turning on/off for an externally started recording. Please add that short live UI check (or have the owner explicitly waive that merge gate).

Non-blocking cleanup while there: move `isRecordingVideo = on` into the same main-queue block as the menu update so notification callbacks cannot write toggle state off-main.

**Comment by liususan091219** [comment]
**E2E menu-indicator live check** (the app-side item from your CR).

Built the PR app and ran it, then posted the `com.sutando.recording.{on,off}` Darwin notification — exactly what the capture server emits on an external start / watchdog auto-stop. The **Drop Video Clip** menu row flips live:

menu indicator E2E — OFF vs ON

- **recording.off** → `Drop Video Clip`
- **recording.on** → `🔴 Drop Video Clip — recording…`

So an externally-started recording drives the app menu indicator (observed on the real app, not just the server), and — per the fix — `isRecordingVideo` is reconciled in the same helper, so ⌃R stops on the first press.

I'll also apply your non-blocking suggestion (move the `isRecordingVideo` write into the same main-queue block as the menu update) in a follow-up commit so the toggle state is never written off-main.

**Review by john-the-dev — APPROVED** [review]
Approved on current head `e2f9011` after the requested live app check closed the last merge gate.

Verified evidence:
- The author built and ran the PR app, posted the same `com.sutando.recording.on/.off` Darwin notifications emitted by the capture server, and attached an observed OFF/ON menu screenshot to the PR. The menu row changed from `Drop Video Clip` to `🔴 Drop Video Clip — recording…`.
- Current code routes those notifications through `setRecordingIndicator`, which updates `isRecordingVideo` on the main queue together with the menu item; the next ⌃R therefore takes the stop path on the first press.
- The `?max=0` race is closed by rejecting non-positive caps and assigning `_active_recording` before starting the watchdog under the recording lock.
- The current cumulative diff has no added hardcoded host paths and `git diff --check origin/main...e2f9011` passes.
- All 16 required GitHub checks and CLA are green; focused results previously verified 6/6 new recording tests and 8/8 existing video-routing tests.

No remaining blockers from my prior REQUEST_CHANGES review.

**Comment by liususan091219** [comment]
Independent verification of the two CR commits (0a093ed7 + e2f9011e) against the review items, from the session that hit both bugs live tonight:

1. **?max<=0 race**: `int(max_raw) > 0` now rejects zero (isdigit already excludes negatives/non-numeric), and `_active_recording` is registered **before** `wd.start()` — with both the request path and `_auto_stop` under `_recording_lock`, an instant-fire timer blocks until registration completes. Race closed by ordering + mutual exclusion, not by timing assumptions. Boundary test present.
2. **External-start state sync**: `isRecordingVideo` is now written in the Darwin-notification observer on the main queue, replacing both direct writes in `dropVideoClip` — toggle state and indicator share one source of truth, which is exactly the double-press repro we hit at 20:08 ET (watchdog auto-stop left the row red while `starting` computed true).

Ops note for whoever merges: tonight's live workaround on the studio machine (env-raised default cap, running since 17:05 ET) becomes obsolete on merge — the post-merge restart per the PR body supersedes it; I'll handle that swap on this end.

(This account authored the PR so it can't self-approve — verification only.)

**Review by qingyun-wu — DISMISSED** [review]
Blocking state race: both the explicit-stop path (`src/screen-capture-server.py:265`) and a stale watchdog callback (`:323-328`) unconditionally publish recording-off after their old process exits. A new recording can start during that finalization and publish on, then the old stop/watchdog publishes off, leaving Sutando.app's `isRecordingVideo` false while the new recording is active. Please gate the off notification on ownership/generation of the same active recording (or serialize stop/start finalization) and add a concurrent stop→start / stale-watchdog regression test.

File: src/screen-capture-server.py [hunk 1]
```diff
@@ -309,19 +328,37 @@ def do_GET(self):
...
+                    wd = threading.Timer(cap, _auto_stop)
                     wd.daemon = True
-                    wd.start()
+                    # Register the active recording BEFORE arming the watchdog: a
+                    # tiny cap could fire _auto_stop almost immediately, and it must
+                    # see _active_recording already set (both run under
+                    # _recording_lock, so the callback blocks until this returns).
                     _active_recording = {"proc": proc, "path": path, "watchdog": wd}
+                    wd.start()
+                _post_recording_state(True)
```
Comment (qingyun-wu) [inline src/screen-capture-server.py]: [P1] Publish `recording.on` while this start still owns the state. The lock is released before this notification, so a concurrent `action=stop` can clear `_active_recording`, finish, and publish OFF; this request then resumes and publishes stale ON even though no recording is active. I reproduced that interleaving deterministically on this head: final `_active_recording is None`, notifications `[False, True]`. Move the ON publish into the ownership-critical section (or re-check the same `proc` under the lock immediately before publishing) and add the mirror start→stop ordering regression test.

**Review by qingyun-wu — DISMISSED** [review]
The new commit fixes the previously reported stale-OFF cases and the focused suites pass (8/8 recording coverage, 8/8 existing video routing, `swiftc -parse`, and `git diff --check`). One blocking mirror race remains in the start path: START releases `_recording_lock` before publishing ON, allowing a completed concurrent STOP to be overwritten by a stale ON notification. Please keep the state mutation and corresponding notification ordered by ownership and cover that interleaving.

**Review by qingyun-wu — DISMISSED** [review]
Re-reviewed at head `acb72a03`. **Finding #1 is fixed; finding #2 (my 2026-07-24 13:51 review) is not — the block stands on that one only.**

**✅ Fixed — the OFF-side race (stop + watchdog).** Both off-publishes are now ownership-gated under `_recording_lock`: the stop path re-checks `_active_recording is None` before `_post_recording_state(False)` (`:269-271`), and `_auto_stop` only publishes off when it still owns the proc (`:335-338`). Both covered by regression tests (`test_stop_does_not_publish_off_when_new_recording_started`, `test_stale_watchdog_does_not_publish_off`). Correct.

**❌ Still blocking — the ON-side mirror race.** The START path publishes ON *outside* the lock:

```
297             with _recording_lock:
...
359                 _active_recording = {"proc": proc, "path": path, "watchdog": wd}
360                 wd.start()
361             _post_recording_state(True)     # <-- indent 16 == same level as `with`; runs AFTER release
```

`_post_recording_state(True)` at `:361` is unconditional and executes after the lock is released. Interleaving that still loses:

1. START-A registers `_active_recording = A` under the lock (`:359`), releases at end of `:360`.
2. STOP arrives: takes the lock, `_active_recording = None` (`:250-252`), releases, finalizes A (SIGINT+wait), re-takes the lock, sees `None` → publishes **OFF** (`:271`). Correct — A is stopped.
3. START-A resumes at `:361` and publishes **ON**. Stale ON stomps the correct OFF → `isRecordingVideo=true` while nothing records.

This is exactly the mirror of the OFF fix, and symmetrical to how you already handle OFF. **One-line fix:** move `_post_recording_state(True)` *inside* the `with _recording_lock:` block (indent to 20, right after `wd.start()`), so the ON notification is atomic with the `_active_recording` mutation — same treatment OFF gets at `:271`/`:338`. Then a concurrent STOP can only publish OFF *after* the atomic (set-A + publish-ON), yielding the correct final OFF.

Please also add the symmetric regression test: START registers A, an interleaved STOP sets `None` + publishes OFF, then assert the final published state is OFF (no stale ON) — the ON-side twin of `test_stop_does_not_publish_off_when_new_recording_started`. The two current tests both exercise the OFF side only.

Everything else looks good — `?max<=0` watchdog guard, the 4h cap, and the main-queue `isRecordingVideo` write from the earlier rounds all hold.

**Review by qingyun-wu — APPROVED** [review]
The start-side mirror race is fixed on `d35b3ae`: recording-ON is now published while `_recording_lock` still owns the newly registered process, so a concurrent stop can only publish OFF afterward. The new regression confirms ON is emitted under the ownership lock. Focused validation passed: 9/9 recording-coverage tests, 8/8 existing video-routing tests, `swiftc -parse`, and `git diff --check`. No remaining blockers from my prior review.

Reviewed by Qingyun's Personal Codex.

**Comment by john-the-dev** [comment]
@cla-assistant check

**Comment by john-the-dev** [comment]
Reopening immediately to retrigger the missing CLA status after update-branch; all required CI and two current-head approvals are green.

**Review by qingyun-wu — APPROVED** [review]
Re-reviewed current head `4a1389e`.

This is a merge-only refresh over the previously reviewed `d35b3ae`; the PR-owned screen-capture files have no content delta from that approved head. The stale OFF/ON mirror-race fixes still hold: state changes and notifications remain ordered under the recording ownership lock, so an old stop/watchdog cannot overwrite a newer recording state and start cannot publish a stale ON after a completed stop.

Focused validation passed:
- `python3 tests/screen-capture-recording-coverage.test.py` (9/9; rerun with loopback socket permission)
- `python3 tests/screen-capture-video-routing.test.py` (8/8; rerun with loopback socket permission)
- `swiftc -parse src/Sutando/main.swift`
- `git diff --check origin/main...HEAD`

Reviewed by Qingyun's Personal Codex.
